language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -193,16 +193,16 @@ dummyKind = Type ()
 dummyType :: Type TyNameWithKind ()
 dummyType = TyVar () dummyTyName
 
--- | Extract type of a term.
-typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (Type TyNameWithKind ())
-typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = pure (void ty)
-typeOf (LamAbs _ _ ty t)                         = TyFun () (void ty) <$> typeOf t
+-- | Extract type of a term. The resulting type is normalized.
+typeOf :: Term TyNameWithKind NameWithType a -> TypeCheckM a (NormalizedType TyNameWithKind ())
+typeOf (Var _ (NameWithType (Name (_, ty) _ _))) = pure (void $ NormalizedType ty) -- annotations on types must be normalized
+typeOf (LamAbs _ _ ty t)                         = NormalizedType <$> (TyFun () (void ty) <$> (getNormalizedType <$> typeOf t))
```

Comment:
There is a lot of ceremony with those `NormalizedType` and `getNormalizedType`. I'm going to propose something strange: what if we make it

```haskell
newtype Normalized a = Normalized
    { unNormalized :: a
    }
```

and add

```haskell
newtype NormalizedT m a = NormalizedT
    { unNormalizedT :: m (Normalized a)
    }
```

Then we can write

```haskell
typeOf :: Term TyNameWithKind NameWithType a -> NormalizedT (TypeCheckM a) (Type TyNameWithKind ())
```

and *maybe* this will be nicer. Just an idea.

Reply:
I'm not wild about this - I think it'll be less obvious what's going on and I've never seen anyone do something like this before to avoid newtype wrapping, so it'll be extra surprising.

Reply:
I agree with what you say, but I'm really curious whether it can improve readability at the term level, so I'm going to try this myself later.

Reply:
I will happily have a look at your experiment ;)